### PR TITLE
1108 barchart component

### LINKED ISSUE #1107: Piechart component
[Something like this, but custom. ](https://mui.com/x/react-charts/pie/)
- A lot of the functionality in the example is probably not needed. Keep it simple, but modular.

Use case:
- recruitment statistics
- other statistics (f.eks. "kronerulling")

### LINKED ISSUE #1108: Barchart component
[Something like this, but custom. ](https://mui.com/x/react-charts/bars/)
- A lot of the functionality in the example is probably not needed. Keep it simple, but modular.

Use case:
- recruitment statistics
- other statistics (f.eks. "kronerulling")

### LINKED ISSUE #1109: Linechart component
[Something like this](https://mui.com/x/react-charts/lines/)

- A lot of the functionality in the example is probably not needed. Keep it simple, but modular.

Use case:
- recruitment statistics
- other statistics (f.eks. "kronerulling")

### PR DESCRIPTION
**Pull request by Snorre98**
This became a big one 😶‍🌫️ In my defense; 12 of the files are below 10 lines and 17 of them below 50 lines. Most of this is types, constants or styling. **9** of the files contain the core functionality. **EXCLUDING COMMENTS**

#### What and why:
This PR creates PieChart, BarChart, and LineChart with a parent Chart component. This can be used to display recruitment statistics or stuff like "Kronerulling".

The PR contains a lot of added files because I have tried to make the components as modular as reasonable possible. What I have implemented can be used to create more chart types. 

‼️ I was not able to solve a type problem in Charts.tsx where TS is unable to recognize the categorized prop. TS gives a warning if for example cartesian chart props are used in a piechart, which is what I was aiming for, so I feel this works well enough.

‼️ For cartesian charts darkmode does not work in Storybook because the theme switching is being done with the useIsDarkeTheme-hook  in the TSX files, which apparently does not update in StoryBook.

Closes #1108 
Closes #1107 
Closes #1109

#### Docs
The props of the Chart component is document in JDoc comments. 
There is some comments explaining stuff in the code.

#### [STORYBOOK LINK](http://localhost:6006/?path=/story/components-chart--applicants-per-day)

*The labels have been updated to stand at 45deg after I took the screenshots*
#### BarChart:

image
image

#### LineChart:

image
image
image

#### PieChart

image
image

### CODE REVIEW

**Review by magsyg — APPROVED** [review]
Sett gjennom ser aight ut, sikkert noe problemmer noe sted her, men de finner vi nok ut av det når dette brukes

File: frontend/src/Components/Chart/CartesianCharts/BarChart/BarChart.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    maxValue,
+    xLabelFreq,
+    svgScale,
+    datapointPadding,
+    leftPadding,
+    bottomPadding,
+    yLabelsPosition,
+    xLabelsMargin,
+    datapointWidth,
+  } = dimensions(sizes, size, data); // function which hold/calculate dimensions or other needed values.
```
Comment (snorrekr) [inline frontend/src/Components/Chart/CartesianCharts/BarChart/BarChart.tsx:38]: Can't this be outside the function?

File: frontend/src/Components/Chart/CartesianCharts/BarChart/index.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+export { BarChart } from './BarChart';
```
Comment (snorrekr) [inline frontend/src/Components/Chart/CartesianCharts/BarChart/index.ts:1]: Nice!

File: frontend/src/Components/Chart/CartesianCharts/BarChart/BarChart.tsx [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+  const lineChartPalette = palette; // imported from utils
+  let colors: CartesianChartsColors;
+
+  isDarkMode ? (colors = lineChartPalette.dark) : (colors = lineChartPalette.light);
```
Comment (snorrekr) [inline frontend/src/Components/Chart/CartesianCharts/BarChart/BarChart.tsx:43]: Can be moved out

File: frontend/src/Components/Chart/CartesianCharts/LineChart/LineChart.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    bottomPadding,
+    yLabelsPosition,
+    xLabelsMargin,
+    datapointWidth,
+  } = dimensions(sizes, size, data); // function which hold/calculate dimensions or other needed values.
+
+  const lineChartPalette = palette; // imported from utils
+  let colors: CartesianChartsColors;
+
+  isDarkMode ? (colors = lineChartPalette.dark) : (colors = lineChartPalette.light);
```
Comment (snorrekr) [inline frontend/src/Components/Chart/CartesianCharts/LineChart/LineChart.tsx:42]: Same here

File: frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    sizes[size];
+
+  const svgWidth = cWith + rightPadding + leftPadding + xLabelsMargin;
+  const svgHeight = cHeight + topPadding + bottomPadding + yLabelsPosition;
+
+  // after adding padding and margin this should be the chart width and height
+  const chartWidth = sizes[size].cWith;
+  const chartHeight = sizes[size].cHeight;
+
+  const maxValue = Math.max(...data.map((item) => item.value));
```
Comment (snorrekr) [inline frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts:27]: Cant you use reducer

Comment (snorrekr) [reply frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts:27]: `data.reduce((acc, curr) => acc + curr.value, 0);`

File: frontend/src/Components/Chart/CartesianCharts/utils/draw-labels.tsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    lines.push(
+      <text
+        key={i}
+        x={yLabelsPosition}
+        y={yPosition + sizes[size].labelFont / 2}
+        style={{ fontSize: sizes[size].labelFont }}
+        fill={colors.text}
+        textAnchor="end"
+      >
+        {splitYLabel ? value.toFixed(0).toString().slice(splitYLabel[0], splitYLabel[1]) : value.toFixed(0)}
```
Comment (snorrekr) [inline frontend/src/Components/Chart/CartesianCharts/utils/draw-labels.tsx:32]: Maybe a comment explaining

File: frontend/src/Components/Chart/CartesianCharts/utils/draw-labels.tsx [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        <text
+          key={index}
+          x={getX(index)}
+          y={svgHeight - xLabelsMargin}
+          style={{ fontSize: sizes[size].labelFont }}
+          fill={colors.text}
+          textAnchor="middle"
+          // Rotate labels for better readability
+          transform={`rotate(-45, ${getX(index)}, ${svgHeight - xLabelsMargin})`}
+        >
```
Comment (snorrekr) [inline frontend/src/Components/Chart/CartesianCharts/utils/draw-labels.tsx:67]: Nice

File: frontend/src/Components/Chart/Chart.tsx [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+ * (+ unit magnitude if relevant).
+ * @param splitYLabel 👉 cartesian only. Allows for splitting the y-axis label. Usually a number value, but can split.
+ * Nice if large number. E.g. '1000000', can be split to '1'.
+ * @param spliceXLabel 👉 cartesian only. Allows for splitting the x-axis label.
+ * Used if the label is long e.g. '24. december', could be split to '24. dec'
+ * @param yLabelCount 👉 cartesian only. Controls amount of labels on the y-axis.
+ * @param hasXDirLines 👉 cartesian only. Controls weather the chart has horizontal lines.
+ * @param hasYDirLines 👉 cartesian only. Controls weather the chart has vertical lines.
+ * @param legend 👉 circle only. description of chart data.
+ * */
```
Comment (snorrekr) [inline frontend/src/Components/Chart/Chart.tsx:37]: :vomiting_face:

File: frontend/src/Components/Chart/CircularCharts/PieChart/PieChart.tsx [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+import { Text } from '~/Components/Text/Text';
+import styles from './PieChart.module.scss';
+import { CircularChartProps } from './utils/types';
+import { sizes } from './utils/apperance';
+
+export function PieChart({ data: initialData, chartTitle, size, legend }: CircularChartProps) {
+  const [dataWithColors, setDataWithColors] = useState<{ color: string; label: string; value: number }[]>([]);
+  const { hoverInfo, handleMouseEnter, handleMouseMove, handleMouseLeave, objectRef } = useHoverLabel();
+  const radius = 200;
+  const viewboxSize = radius * 2;
```
Comment (snorrekr) [inline frontend/src/Components/Chart/CircularCharts/PieChart/PieChart.tsx:14]: Move outside

**Review by snorrekr — APPROVED** [review]
nice, some comments added.
I think its fine to add instead of a library. While a library can achieve the same as the PR, with more functionality, this PR is implemented so why waste time implementing another solution. Additionally we dont have many charts available to the public, if that becomes a major requirment it can be added then.

File: frontend/src/Components/Chart/CartesianCharts/CartesianCharts.module.scss [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+.vLegendContainer{
+  grid-area: vLegend;
+  @include flex-column-center;
+  width: 1rem;
+}
+
+.vAxisLegend{
+  rotate: -90deg;
```
Comment (robines) [inline frontend/src/Components/Chart/CartesianCharts/CartesianCharts.module.scss:31]: Ser ut til at teksten clippes av roteringen
image

File: frontend/src/Components/Chart/CartesianCharts/BarChart/index.ts (hunk 2 above)
Comment (robines) [reply frontend/src/Components/Chart/CartesianCharts/BarChart/index.ts:1]: @snorrekr resolve? 🤡

**Review by robines — CHANGES_REQUESTED** [review]

File: frontend/src/Components/Chart/CartesianCharts/utils/types.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+  yLabelCount: number;
+  hasXDirLines?: boolean;
+  hasYDirLines?: boolean;
+  splitYLabel?: [number, number];
+  splitXLabel?: [number, number];
+};
+
+export type CartesianChartsSize = {
+  cWith: number;
+  cHeight: number;
```
Comment (robines) [inline frontend/src/Components/Chart/CartesianCharts/utils/types.ts:21]: Her må hjernen jobbe ekstra for å skjønne hva `c` er. Syns det er mye tydeligere å bare ha `width`/`height`

```suggestion
  width: number;
  height: number;
```

File: frontend/src/Components/Chart/CartesianCharts/utils/types.ts [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  small: CartesianChartsSize;
+  medium: CartesianChartsSize;
+  large: CartesianChartsSize;
+  xlarge: CartesianChartsSize;
+};
+
+export type CartesianChartsColors = {
+  bar: string;
+  bg: string;
+  text: string;
```
Comment (robines) [inline frontend/src/Components/Chart/CartesianCharts/utils/types.ts:41]: Syns det er tydeligere med dette

```suggestion
  background: string;
  foreground: string;
```

File: frontend/src/Components/Chart/CartesianCharts/CartesianCharts.module.scss (hunk 10 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/CartesianCharts/CartesianCharts.module.scss:31]: Ye, har funksjonalitet for å kontrollere lengden på label stringen, men kan se på å utvide svg-en dynamisk

File: frontend/src/Components/Chart/Chart.tsx [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+export function Chart({
```
Comment (robines) [inline frontend/src/Components/Chart/Chart.tsx:38]: Jeg skjønner ikke helt vitsen med denne komponenten, den virker bare overflødig? Mye tydeligere å bruke enten `BarChart` eller `LineChart` direkte i koden.

File: frontend/src/Components/Chart/CartesianCharts/BarChart/BarChart.tsx (hunk 3 above)
Comment (robines) [reply frontend/src/Components/Chart/CartesianCharts/BarChart/BarChart.tsx:43]: Ikke ha assignments i ternary expressions, det er megaforvirrende

Comment (robines) [reply frontend/src/Components/Chart/CartesianCharts/BarChart/BarChart.tsx:43]:
```suggestion
  const colors = isDarkMode ? palette.dark : palette.light;
```

File: frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts [hunk 14]
```diff
@@ -0,0 +1,74 @@
+import {
+  CartesianChartProps,
+  CartesianChartsData,
+  CartesianChartSizes,
+} from '~/Components/Chart/CartesianCharts/utils/types';
+
+/*
+ * Determines the spacing between bars. 1px if there are more than 80 entries, 2px if less.
+ * */
+const getDatapointPadding = (entryCount: number) => (entryCount > 80 ? 1 : 2);
```
Comment (robines) [inline frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts:10]:
```suggestion
function getDatapointPadding(entryCount: number) {
  return entryCount > 80 ? 1 : 2;
}
```

File: frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+const calculateDimensions = (
+  sizes: CartesianChartSizes,
+  size: CartesianChartProps['size'],
+  data: CartesianChartsData[],
+) => {
```
Comment (robines) [inline frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts:16]:
```suggestion
function calculateDimensions(
  sizes: CartesianChartSizes,
  size: CartesianChartProps['size'],
  data: CartesianChartsData[],
) {
```

File: frontend/src/_constants.scss [hunk 16]
```diff
@@ -6,6 +6,8 @@ $red-samf-faded: #fac7c8;
 $red-lighter: #fff2f0;
 $red-light: #ffcfca;
 $red: #dc1010;
+$salmon: #fa8072;
+$salmon_light: #fab4ac;
```
Comment (robines) [inline frontend/src/_constants.scss:10]: Follow naming convention

```suggestion
$salmon-light: #fab4ac;
```

**Review by robines — CHANGES_REQUESTED** [review]

File: frontend/src/Components/Chart/CartesianCharts/LineChart/LineChart.tsx (hunk 4 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/CartesianCharts/LineChart/LineChart.tsx:42]: Nei, kan ikke flytte ut. Bruker en hook.

File: frontend/src/Components/Chart/CartesianCharts/utils/draw-labels.tsx (hunk 7 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/CartesianCharts/utils/draw-labels.tsx:67]: ok

File: frontend/src/Components/Chart/Chart.tsx (hunk 8 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/Chart.tsx:37]: Ok

File: frontend/src/Components/Chart/CircularCharts/PieChart/PieChart.tsx (hunk 9 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/CircularCharts/PieChart/PieChart.tsx:14]: Ok

File: frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts (hunk 14 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts:10]: ok

File: frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts (hunk 15 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts:16]: ok

File: frontend/src/Components/Chart/CartesianCharts/BarChart/BarChart.tsx (hunk 3 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/CartesianCharts/BarChart/BarChart.tsx]: fixed

File: frontend/src/Components/Chart/Chart.tsx (hunk 8 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/Chart.tsx]: fixed

File: frontend/src/Components/Chart/CircularCharts/PieChart/PieChart.tsx (hunk 9 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/CircularCharts/PieChart/PieChart.tsx]: fixed

File: frontend/src/_constants.scss (hunk 16 above)
Comment (Snorre98) [reply frontend/src/_constants.scss]: fixed

File: frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts (hunk 14 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts]: fix

File: frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts (hunk 15 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/CartesianCharts/utils/dimensions.ts]: fixed

File: frontend/src/Components/Chart/Chart.tsx (hunk 13 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/Chart.tsx:38]: nei

File: frontend/src/Components/Chart/CartesianCharts/BarChart/BarChart.tsx (hunk 3 above)
Comment (snorrekr) [reply frontend/src/Components/Chart/CartesianCharts/BarChart/BarChart.tsx]: Mer som ut av funksjoner.

File: frontend/src/Components/Chart/Chart.tsx (hunk 13 above)
Comment (Snorre98) [reply frontend/src/Components/Chart/Chart.tsx:38]: @robines er PR-en good nå eller er det er dette dealbreaker for deg?

Comment (robines) [reply frontend/src/Components/Chart/Chart.tsx:38]: @Snorre98 Ikke dealbreaker, selv om jeg fremdeles ikke ser poenget.

**Review by robines — APPROVED** [review]

**Pull request merged**